torchgen/gen_functionalization_type.py

```diff
@@ -363,7 +363,17 @@ def emit_view_functionalization_body(
       );
       {return_type} reference_tensor_output;
       {{
-        at::AutoDispatchSkipFunctionalize guard;
+        // Exclude any modes: the purpose of calling into meta kernels is only as an implementation
+        // detail to perform shape inference, and we don't want any modal keys to run.
+        // Specifically, we want to prevent functionalization and Python modes from running.
+        c10::impl::ExcludeDispatchKeyGuard guard(
+            c10::functorch_transforms_ks |
+            c10::DispatchKeySet({{
+              c10::DispatchKey::FuncTorchDynamicLayerBackMode,
+              c10::DispatchKey::FuncTorchDynamicLayerFrontMode,
+              c10::DispatchKey::Python
```

Comment:
FWIW, I originally thought it would be simpler just to "exclude all keys in TLS", so we're guaranteed to hit the (meta) backend. That breaks though because we still need BackendSelect. Instead of hardcoding the valid set (something like "exclude all keys in TLS, but re-enable the default TLS), I opted to just exclude the keys that I know will break things:
- `Python`, because of `ProxyTorchDispatchMode`
- `FuncTorch` layer mode keys, because if we redispatch to one of them, then they can trample on the existing TLS that I set

If anyone has a strong opinion about a different way to set up the TLS I'm all ears :)

Reply:
So, if Python is excluded, that means that if a Meta implementation is written from Python we will not see it, is that right?

Reply:
Maybe you should remind me why we need to do a full dispatch here, instead of just using our torchgen powers to find the true underlying native:: kernel and calling it directly haha.

Reply:
> Maybe you should remind me why we need to do a full dispatch here, instead of just using our torchgen powers to find the true underlying native:: kernel and calling it directly haha.

We could call the underlying `native::` kernels, but many of them are composite, so we end up with the same problem - they'll redispatch to other operators and look at TLS, so we need to zero it out.

> So, if Python is excluded, that means that if a Meta implementation is written from Python we will not see it, is that right?

Hmmm I don't think so? the registration-from-python-logic registers those functions directly to the `Meta` dispatch key, IIRC. So the Python key is still reserved only for python subclasses / python modes.

Reply:
can't we call into `at::meta::func()` here?

Reply:
> We can't conveniently call at::meta::func(), since some ops are CompositeExplicitAutograd.

Ho the alias keys also get namespace? Not sure why but I was expecting to have namespaces only for keys and would get the functions there.

In any case, with the increase of dynamic addition to the dispatcher (via python registration) I guess we want to discourage using these namespaces as much as possible.

Reply:
> Ho the alias keys also get namespace? 

For better or for worse, today yep :p I guess we could figure out how to change this at some point, it might make LTC codegen a bit cleaner too